interpreter/beam/beam.go

```diff
@@ -0,0 +1,510 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package beam // import "go.opentelemetry.io/ebpf-profiler/interpreter/beam"
+
+// BEAM VM Unwinder support code
+
+// The BEAM VM is an interpreter for Erlang, as well as several other languages
+// that share the same bytecode, such as Elixir and Gleam.
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+	"unsafe"
+
+	log "github.com/sirupsen/logrus"
+
+	"go.opentelemetry.io/ebpf-profiler/host"
+	"go.opentelemetry.io/ebpf-profiler/interpreter"
+	"go.opentelemetry.io/ebpf-profiler/libpf"
+	"go.opentelemetry.io/ebpf-profiler/libpf/pfelf"
+	"go.opentelemetry.io/ebpf-profiler/lpm"
+	"go.opentelemetry.io/ebpf-profiler/process"
+	"go.opentelemetry.io/ebpf-profiler/remotememory"
+	"go.opentelemetry.io/ebpf-profiler/reporter"
+	"go.opentelemetry.io/ebpf-profiler/support"
+)
+
+// #include "../../support/ebpf/types.h"
+// #include "../../support/ebpf/v8_tracer.h"
+import "C"
+
+var (
+	// regex for matching the process name
+	beamRegex                      = regexp.MustCompile(`beam.smp`)
+	_         interpreter.Data     = &beamData{}
+	_         interpreter.Instance = &beamInstance{}
+)
+
+type beamData struct {
+	version               uint32
+	the_active_code_index uint64
+	r                     uint64
+	erts_atom_table       uint64
+}
+
+type beamInstance struct {
+	interpreter.InstanceStubs
+
+	pid       libpf.PID
+	data      *beamData
+	rm        remotememory.RemoteMemory
+	rangesPtr libpf.Address
+	atomTable libpf.Address
+}
+
+func readSymbolValue(ef *pfelf.File, name libpf.SymbolName) ([]byte, error) {
```

Comment:
fyi, if you rebase, there is now `pfelf.File.SymbolData` that returns the symbol info + relevant code portion. This is superior as the code return can be mmapped data meaning no copies needed.